src/utils/scrap_data.ts

```diff
@@ -0,0 +1,176 @@
+    return wineStylesResponse?.wine_styles;
+  };
+
+  getFoods = async (page: Page): Promise<Food[]> => {
+    const wineStylesResponse = await page.evaluate(async apiUrl => {
+      const cacheKey = localStorage.getItem('local_cache_key') + window['vivinoCacheKey'];
+      return fetch(`${apiUrl}?cache_key=${cacheKey}`).then((res: any) => {
+        return res.text().then((r: any) => JSON.parse(r));
+      });
+    }, VIVINO_FOODS_API);
+    return wineStylesResponse?.foods;
+  };
+
+  getAllWineListResponse = async (page: Page, wineParams: WineParams): Promise<Wine[]> => {
+    const winesList: Wine[] = [];
+    const firstPageResponse = await this.getWinesListResponsePerPage(page, 1, wineParams);
+    const totalRecords = firstPageResponse?.records_matched;
+    winesList.push(...this.mapWinesToWineObjects(firstPageResponse?.matches));
+    const totalPagesToFetch = Math.ceil(totalRecords / WINES_PER_PAGE);
+    console.log('total records', totalRecords);
+    console.log('total pages to fetch', totalPagesToFetch);
+    for (let pageIndex = 2; pageIndex <= totalPagesToFetch; pageIndex++) {
+      const response = await this.getWinesListResponsePerPage(page, pageIndex, wineParams);
+      winesList.push(...this.mapWinesToWineObjects(response?.matches));
+    }
+    return winesList;
+  };
+
+  getWinesListResponsePerPage = async (page: Page, pageIndex: number, wineParams: WineParams): Promise<any> => {
+    const url = this.getVivinoSearchUrl(pageIndex, wineParams);
+    const grapesResponse = await page.evaluate(async url => {
+      return fetch(url).then((res: any) => {
+        return res.text().then((r: any) => JSON.parse(r));
+      });
+    }, url);
+    return grapesResponse?.explore_vintage;
+  };
+
+  public getWinesDataFromVivino = async (wineParams: WineParams): Promise<Wine[]> => {
+    try {
+      const browser: Browser = await puppeteer.launch({ headless: false, timeout: 100000000 });
+      const page: Page = await browser.newPage();
+
+      // setting user agent because puppeteer fails when setting headerless true
+      await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36');
+      // disable loading images in the browser to speed up execution
+      await page.setRequestInterception(true);
+      page.on('request', request => {
+        if (request.resourceType() === 'image') request.abort();
+        else request.continue();
+      });
+      await page.goto(`${VIVINO_SEARCH_URL}`);
+      await page.waitForSelector('.simpleLabel__selectedKey--11QuD');
+      await page.click('.simpleLabel__selectedKey--11QuD');
+      await page.click('.shipToDropdown__list--1_3yJ li:last-child a');
+      await page.waitForNavigation({ waitUntil: 'networkidle0' });
+      this.grapes = await this.getGrapes(page);
+      if (!this.isGrapeIsValid(wineParams.grapes)) {
+        await browser.close();
+        throw 'Grape is invalid';
```

Comment:
You have a HttpException class but here you just using a string. Why?

Reply:
Since the requirement of the exercise is to create a project that can be run from command line, I thought it is better to return as string for better readability from the command line.